algebra/euclidean_domain.lean

```diff
@@ -0,0 +1,298 @@
+  simp at this,
+  exact this,
+end
+
+lemma dvd_mod_self (a : α) : a ∣ a % a :=
+begin
+  let d := (a/a)*a, -- ring tactic can't solve things without this
+  have : a%a = a - (a/a)*a, from
+    calc
+      a%a = d + a%a  - d : by ring
+      ... = (a/a)*a + a%a - (a/a)*a : by dsimp [d]; refl
+      ... = a - (a/a)*a : by simp [(%), (/), quotient_mul_add_remainder_eq a a],
+  rw this,
+  exact dvd_sub (dvd_refl a) (dvd_mul_left a (a/a)),
+end
+
+lemma mod_lt  : ∀ (a : α) {b : α}, valuation b > valuation (0:α) →  valuation (a%b) < valuation b :=
+begin
+  intros a b h,
+  by_cases b_zero : (b=0),
+  {
+    rw b_zero at h,
+    have := lt_irrefl (valuation (0:α)),
+    contradiction,
+  },
+  {
+    exact valuation_remainder_lt a b b_zero,
+  }
+end
+
+lemma neq_zero_lt_mod_lt (a b : α) :  b ≠ 0 → valuation (a%b) < valuation b
+| hnb := valuation_remainder_lt a b hnb
+
+lemma dvd_mod {a b c : α} : c ∣ a → c ∣ b → c ∣ a % b :=
+begin
+  intros dvd_a dvd_b,
+  have : remainder a b = a - quotient a b * b, from
+  calc 
+    a%b = quotient a b * b + a%b - quotient a b * b : by ring
+    ... = a - quotient a b * b : by {dsimp[(%)]; rw (quotient_mul_add_remainder_eq a b)},
+  dsimp [(%)], 
+  rw this,
+  exact dvd_sub dvd_a (dvd_mul_of_dvd_right dvd_b (a/b)),
+end
+
+/- strong lemmas -/
+
+lemma val_lt_one (a : α) : valuation a < valuation (1:α) → a = 0 :=
+begin
+  intro a_lt,
+  by_cases a = 0,
+  { exact h},
+  { 
+    have := le_valuation_mul (1:α) a h,
+    simp at this,
+    have := not_le_of_lt a_lt,
+    contradiction
+  }
+end
+
+lemma val_dvd_le : ∀ a b : α, b ∣ a → a ≠ 0 → valuation b ≤ valuation a
+| _ b ⟨d, rfl⟩ ha :=
+begin
+  by_cases d = 0,
+  {
+    simp[h] at ha,
+    contradiction
+  },
+  {
+    exact le_valuation_mul b d h,
+  }
+end
+
+@[simp] lemma mod_one (a : α) : a % 1 = 0 := 
+val_lt_one _ (valuation_remainder_lt a 1 one_ne_zero)
+
+@[simp] lemma zero_mod (b : α) : 0 % b = 0 :=
+begin
+  have h : remainder 0 b = b * (-quotient 0 b ), from
+    calc
+      remainder 0 b = quotient 0 b * b + remainder 0 b + b * (-quotient 0 b ) : by ring
+      ...                       = b * (-quotient 0 b ) : by rw [quotient_mul_add_remainder_eq 0 b, zero_add],
+  by_cases quotient_zero : (-quotient 0 b) = 0,
+  {
+    simp[quotient_zero] at h,
+    exact h
+  },
+  {
+    by_cases h'' : b = 0,
+    {
+      rw h'',
+      simp
+    },
+    {
+      have := not_le_of_lt (valuation_remainder_lt 0 b h''),
+      rw h at this,
+      have := le_valuation_mul b (-quotient 0 b) quotient_zero,
+      contradiction,
+    }
+  }
+end
+
+@[simp] lemma zero_div (b : α) (hnb : b ≠ 0) : 0 / b = 0 :=
+begin
+  have h1 : remainder 0 b = 0, from zero_mod b,
+  have h2 := quotient_mul_add_remainder_eq 0 b,
+  simp[h1] at h2,
+  cases eq_zero_or_eq_zero_of_mul_eq_zero h2,
+  { exact h },
+  { contradiction }
+end
+
+@[simp] lemma mod_self (a : α) : a % a = 0 :=
+begin
+  have := quotient_mul_add_remainder_eq a a,
+  cases (dvd_mod_self a) with m a_mul,
```

Comment:
Rather than using cases it makes more sense to write something like `let ⟨m, a_mul⟩ := (dvd_mod_self a)`.
I'm not sure how the syntax for this works in tactics mode though.

Reply:
Why not write the first part of the proof in term mode, and switch to tactic mode when the going gets tough? This still seems to compile:

```
@[simp] lemma mod_self (a : α) : a % a = 0 :=
let ⟨m, a_mul⟩ := dvd_mod_self a in begin
  have := quotient_mul_add_remainder_eq a a,
  by_cases m = 0,
  {

...
```

and indeed you even seem to be able to comment out the `have := quotient_mul_add_remainder_eq a a` line. 

I had a fear of writing proofs in term mode for a long time, but now I realise that actually for some stuff term mode is more natural than tactic mode. Tactic mode comes into its own when you're actually using tactics (e.g. `rw` is far less painful than all the `eq.subst` stuff that term mode would make you do) but sometimes term mode is more natural if you're not using too many tactics.

Reply:
Fixed. I'm not quite sure if my indentation style is correct with the `let` there though.

I have a bad habbit of doing everything in tactics mode.

Reply:
I don't think we should put the `begin` at the end of the previous line. Usually we have the rule that parenthesis should be in the line where the term starts. For me only infix / mixfix operators should stay on the previous line.